apollo-router/src/axum_factory/axum_http_server_factory.rs

```diff
@@ -433,14 +434,37 @@ where
         )
         .layer(middleware::from_fn(metrics_handler));
 
+    if let Some(main_endpoint_layer) = ENDPOINT_CALLBACK.get() {
+        main_route = main_endpoint_layer(main_route);
+    }
+
     let route = endpoints_on_main_listener
         .into_iter()
-        .fold(main_route, |acc, r| acc.merge(r.into_router()));
+        .fold(main_route, |acc, r| {
+            let mut router = r.into_router();
+            if let Some(main_endpoint_layer) = ENDPOINT_CALLBACK.get() {
+                router = main_endpoint_layer(router);
+            }
+
+            acc.merge(router)
+        });
 
     let listener = configuration.supergraph.listen.clone();
     Ok(ListenAddrAndRouter(listener, route))
 }
 
+static ENDPOINT_CALLBACK: OnceLock<Arc<dyn Fn(Router) -> Router + Send + Sync>> = OnceLock::new();
+
+/// Set a callback that may wrap or mutate `axum::Router` as they are added to the main router.
+/// Although part of the public API, this is not intended for use by end users, and may change at any time.
+#[doc(hidden)]
+pub fn set_axum_router_callback(callback: impl Fn(Router) -> Router + Send + Sync + 'static) {
```

Comment:
can we add `unchecked` or `unsafe` to the function name? I think we should also document it may panic if called twice, etc...

Reply:
[6fd7990](https://github.com/apollographql/router/pull/4547/commits/6fd79909e40548c9b81b61ce73ded2de526362e4)